sycl/doc/Assert.md

```diff
@@ -0,0 +1,138 @@
+asynchronous exception handler is set the failure is reported with
+`cl::sycl::event_error` exception. Otherwise, SYCL Runtime should trigger abort.
+At least one failed assertion should be reported.
+
+When multiple kernels are enqueued and both fail at assertion at least single
+assertion should be reported.
+
+## User requirements
+
+From user's point of view there are the following requirements:
+
+| # | Title | Description | Importance |
+| - | ----- | ----------- | ---------- |
+| 1 | Handle assertion failure | Signal about assertion failure via SYCL asynchronous exception | Must have |
+| 2 | Print assert message | Assert function should print message to stderr at host | Must have |
+| 3 | Stop under debugger | When debugger is attached, break at assertion point | Highly desired |
+| 4 | Reliability | Assert failure should be reported regardless of kernel deadlock | Highly desired |
+
+## Contents of `cl::sycl::event_error`
+
+`cl::sycl::event_error::what()` should return the same assertion failure message
+as is printed at the time being.
+
+Other than that, interface of `cl::sycl::event_error` should look like:
+```
+class event_error : public runtime_error {
+public:
+  event_error() = default;
+
+  event_error(const char *Msg, cl_int Err)
+      : event_error(string_class(Msg), Err) {}
+
+  event_error(const string_class &Msg, cl_int Err) : runtime_error(Msg, Err) {}
+};
+```
+
+Regardless of whether asynchronous exception handler is set or not, there's an
+action to be performed by SYCL Runtime. To achieve this, information about
+assert failure should be propagated from device-side to SYCL Runtime. This
+should be performed via calls to `clGetEventInfo` for OpenCL backend and
+`zeEventQueryStatus` for Level-Zero backend.
+
+## Terms
+
+ - Device-side Runtime - part of device-code, which is supplied by Device-side
+   Compiler.
+ - Low-level Runtime - the backend/runtime, behind DPCPP Runtime.
+ - Device-side Compiler - compiler which generates device-native bitcode based
+   on input SPIR-V image.
+ - Accessor metadata - parts of accessor representation at device-side: pointer,
+   ranges, offset.
+
+## How it works?
+
+For the time being, `assert(expr)` macro ends up in call to
+`__devicelib_assert_fail` function. This function is part of [Device library extension](doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst#cl_intel_devicelib_cassert).
+Device code already contains call to the function. Currently, a device-binary
+is always linked against fallback implementation.
+Device-side compiler/linker provides their implementation of `__devicelib_assert_fail`
+and prefer this implementation over fallback one.
+
+If Device-side Runtime supports `__devicelib_assert_fail` then Low-Level Runtime
+is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+When detected, Low-level Runtime reports assert failure to DPCPP Runtime
+at synchronization points.
```

Comment:
According to the example above, the exception `what()` string should contain the assertion message.  How will this work in the case  where the low-level runtime supports `__devicelib_assert_fail` natively?  I don't see any way in the OpenCL or Level Zero extensions for the DPC++ runtime to get the assertion message string.

Reply:
Oh, that's a mis-copied part from previous version of the document.
I will fix it.